### DISCUSSION TITLE
V7 - Removed and unregistered fields added via "useFieldArray" are submitted. Controller's Defaultvalues are submitted instead of input value.

### DISCUSSION THREAD

**Discussion by deniercounter** [react-hook-form/react-hook-form#4515] [Q&A]
Dear community, I have two problems in context of "Controller" with "useFieldArray" that arose with migration from v6 to v7. I lost hours to find a solution to **these two problems**, which seems to be a result of my personal misunderstanding of the v7 way of doing things:

**Description of the 2 problems.**

1) A removed and unregistered form value gets submitted (see all answers of the first question in the console.table log in the CSB: https://codesandbox.io/s/festive-sinoussi-urfsp?file=/src/components/ccf-client/components/CCFClient7AnswerChoices/index.tsx)
2) A manually changed form value shows in the form as changed, nevertheless its defaultValue set as Controller attribute is the submitted value.

**My expectation would have been:**

1) If the "Remove question"-Button (invoking "remove" from "useFieldArray") of an answer is clicked this answer should disappear from the array and not be submitted.
2) If an input field is manually changed to another value, the new value should become the current value that gets submitted.

**Reproduction of problem 1)**
a) Go to: https://codesandbox.io/s/festive-sinoussi-urfsp?file=/src/components/ccf-client/components/CCFClient7AnswerChoices/index.tsx
b) Click on grey Button "Add question"
c) Click on red Button "Add answer"
d) Click a second time on "Add answer"
e) Submit the form
f) Open the console and take a look at the console.table "SUBMIT RESULTS FOR FIRST (!) QUESTION" -> All fine
g) Click on one of the blue Buttons "Remove + Unregister answer @index ..."
h) Submit again
i) Have a look for this submit in the console: **The same values are submitted from the form ... Why not the altered array of answers?**

**Reproduction of problem 2)**
a) Go to: https://codesandbox.io/s/festive-sinoussi-urfsp?file=/src/components/ccf-client/components/CCFClient7AnswerChoices/index.tsx
b) Click on grey Button "Add question"
c) Click on red Button "Add answer"
d) Submit the form
e) Open the console and take a look at the console.table -> All fine
f) Change the question field
g) Submit again
h) Have a look for this submit in the console "SUBMIT RESULTS FOR FIRST (!) QUESTION": The same values are submitted from the form ... **Why has the change of the question been ignored and instead the defaultValue of the Controller been submitted?**

**Used versions**
    "jquery": "3.5.1",
    "popper.js": "1.16.1",
    "react": "17.0.2",
    "react-bootstrap": "^1.5.2",
    "react-dom": "17.0.2",
    "react-hook-form": "7.0.0-rc.7",
    "react-scripts": "4.0.0",
    "bootstrap": "^4.6.0"

**Comment by deniercounter** [comment]
Changing the input from
```
<input
              defaultValue={`Answer ${subanswerIndex + 1} ${subanswer.id}`}
              type="hidden"
              {...register(
                `formvalues.${questionNumber}.subanswers.${subanswerIndex}.answer` as const
              )}
/>
<Controller
              render={({ field }) => <input {...field} className="form-control" />}
              name={`formvalues.${questionNumber}.subanswers.${subanswerIndex}.answer`}
              control={control}
              defaultValue={`Answer ${subanswerIndex + 1} ${subanswer.id}`}
/>
```

to only

```
<input
                {...register(
                  `formvalues.${questionNumber}.subanswers.${subanswerIndex}.newcondition.name` as const
                )}
                type="text"
                defaultValue={`Enter new variable name...${subanswer.id}`}
                className="form-control"
/>
```

resolves Problem 2 and submits the changed input value, but has the negative side effect, that...
```
onClick={() => {
              unregister([
                `formvalues.${questionNumber}.subanswers.${subanswerIndex}.answer`,
                `formvalues.${questionNumber}.subanswers.${subanswerIndex}.newcondition.name`,
                `formvalues.${questionNumber}.subanswers.${subanswerIndex}.newcondition.value`
              ]);
              remove(subanswerIndex);
            }}
```
always removes the last index from "fields.map" returned Array. Deleting Answer 1 results in deleting the last added Answer-set. And in addition even all answers removed, submitting results in also sending the "deleted" AND "removed" answers.

Here the CSB: https://codesandbox.io/s/suspicious-architecture-vwztn?file=/src/components/ccf-client/components/CCFClient7AnswerChoices/index.tsx

Does someone has any idea? A helping hand would be appreciated.

**Comment by bluebill1049** [comment]
why do you need to use 

```
unregister([
    `formvalues.${questionNumber}.subanswers.${subanswerIndex}.answer`,
    `formvalues.${questionNumber}.subanswers.${subanswerIndex}.newcondition.name`,
    `formvalues.${questionNumber}.subanswers.${subanswerIndex}.newcondition.value`
  ]);
 ```
 
`remove` should remove inputs from the field array.

**Reply by deniercounter** [reply]
Because the "remove" from array method didn't actually remove the answers from being submitted, although the component unmounted at the correct index.

Even after removing all previously added answers, all these removed answers were still submitted. I thought that they would still be registered - also because when one adds a new answer after one has been removed, the "old" removed index values re-appeared.

So I hoped by adding an explicit "unregister" within the map "loop" would do the trick. 

But unfortunately it didn't .... and commenting it out didn't change this behavior.

I changed to...
```
onClick={() => {
              // unregister([
              //   `formvalues.${questionNumber}.subanswers.${subanswerIndex}.answer`,
              //   `formvalues.${questionNumber}.subanswers.${subanswerIndex}.newcondition.name`,
              //   `formvalues.${questionNumber}.subanswers.${subanswerIndex}.newcondition.value`
              // ]);
              remove(subanswerIndex);
            }}
```

on https://codesandbox.io/s/suspicious-architecture-vwztn?file=/src/components/ccf-client/components/CCFClient7AnswerChoices/index.tsx -> **but didn't change my strange results**
